lib/driver-cli.js

```diff
@@ -0,0 +1,265 @@
+    stopSpinner();
+    driverData = getDriverFields(pkgJsonData);
+  } else if (github) {
+    installType = INSTALL_TYPE_GITHUB;
+    if (installSpec.split('/').length !== 2) {
+      throw new Error(`Github driver spec ${installSpec} appeared to be invalid; ` +
+                      'it should be of the form <org>/<repo>');
+    }
+    driverData = await installViaNpm(json, installSpec);
+  } else if (git) {
+    installType = INSTALL_TYPE_GIT;
+    driverData = await installViaNpm(json, installSpec);
+  } else {
+    installType = INSTALL_TYPE_NPM;
+
+    // drivers installed via npm can include versions or tags after the '@'
+    // sign, so check for that
+    const [name, pkgVer] = installSpec.split('@');
+    let pkgName;
+
+    if (npm) {
+      // if we're installing a named package from npm, we don't need to check
+      // against the appium driver list; just use the installSpec as is
+      pkgName = name;
+    } else {
+      // if we're installing a named appium driver (like 'xcuitest') we need to
+      // dereference the actual npm package ('appiupm-xcuitest-driver'), so
+      // check it exists and get the correct package
+      const knownNames = Object.keys(KNOWN_DRIVERS);
+      if (!_.includes(knownNames, name)) {
+        const msg = `Could not resolve driver; are you sure it's in the list ` +
+                    `of supported drivers? ${JSON.stringify(knownNames)}`;
+        throw new Error(msg);
+      }
+      pkgName = KNOWN_DRIVERS[name];
+    }
+
+    driverData = await installViaNpm(json, driver, pkgName, pkgVer);
+  }
+
+  if (!driverData.driverName || !driverData.version || !driverData.automationName ||
+      !driverData.platformNames || !driverData.mainClass) {
+    throw new Error(`Installed driver did not expose correct fields for ` +
+                    `compability with Appium.`);
+  }
+
+  const {driverName} = driverData;
+  delete driverData.driverName;
+
+  if (_.includes(Object.keys(INSTALLED_DRIVERS), driverName)) {
+    throw new Error(`A driver named '${driverName}' is already installed. ` +
+                    `Did you mean to update? 'appium driver update'. See ` +
+                    `installed drivers with 'appium driver list --installed'.`);
+  }
+
+  driverData.installType = installType;
+  driverData.installSpec = installSpec;
+  await writeInstalledDriver(driverName, driverData);
+
+  // refresh the installed driver manifest
+  INSTALLED_DRIVERS = validateDriversJson();
+
+  // log info for the user
+  log(json, `Driver ${driverName}@${driverData.version} successfully installed`.green);
+  log(json, `- automationName: ${driverData.automationName.green}`);
+  log(json, `- platformNames: ${JSON.stringify(driverData.platformNames).green}`);
+
+  return INSTALLED_DRIVERS;
+}
+
+async function installViaNpm (json, driver, pkgName, pkgVer) {
+  if (!pkgName) {
+    pkgName = driver;
+  }
+  const npmSpec = `${pkgName}${pkgVer ? '@' + pkgVer : ''}`;
```

Comment:
npm also supports special identifiers, like `latest` or `beta`. Are they properly handled?

Reply:
Yes, any npm identifier will work here. We just pass it straight from the command line.